remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,549 @@
+        return VisitResult::NoProgress;
+      } else {
+        auto offset = (plus ? ptr1->second + ptr2->second : ptr1->second - ptr2->second);
+        offset_map.insert({&I, offset});
+        LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+        return VisitResult::Progress;
+      }
+    }
+  }
+}
+
+// Visit a PHI node and update the offset map.
+VisitResult ForwardAliasVisitor::visitPHINode(llvm::PHINode &I) {
+  LOG(INFO) << "Entered PHI node";
+  // iterate over each operand
+  auto in_offset_map = false;
+  auto in_exclude_set = false;
+  uint64_t offset;
+  for (auto &operand : I.operands()) {
+    if (exclude.count(operand)) {
+      in_exclude_set = true;
+    } else {
+      auto ptr = offset_map.find(operand);
+      if (ptr == offset_map.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        if (!in_offset_map) {
+          offset = ptr->second;
+          in_offset_map = true;
+        } else {
+          if (ptr->second != offset) {
+            // bail if the offsets don't match
+            return VisitResult::Error;
+          }
+        }
+      }
+    }
+  }
+  if (in_offset_map && in_exclude_set) {
+    // fail if some operands are excluded and others are state offsets
+    return VisitResult::Error;
+  } else if (in_offset_map) {
+    LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+    offset_map.insert({&I, offset});
+  } else if (in_exclude_set) {
+    LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+    exclude.insert(&I);
+  }
+  return VisitResult::Progress;
+}
+
+// Return the scope of the given instruction.
+llvm::MDNode *GetScopeFromInst(llvm::Instruction &I) {
+  llvm::AAMDNodes N;
+  I.getAAMetadata(N);
+  return N.Scope;
+}
+
+// Remove all dead stores.
+void RemoveDeadStores(const std::unordered_set<llvm::Instruction *> &dead_stores) {
+  for (auto *store : dead_stores) {
+    store->eraseFromParent();
+  }
+}
+
+LiveSetBlockVisitor::LiveSetBlockVisitor(const ScopeMap &scope_to_offset_)
+  : scope_to_offset(scope_to_offset_),
+    curr_wl(),
+    next_wl(),
+    block_map(),
+    to_remove() {}
+
+// Get every terminating basic block from the function `func`.
+void LiveSetBlockVisitor::AddFunction(llvm::Function &func) {
+  for (auto &block : func) {
+    // The machines rose from the ashes of the nuclear fire....
+    if (block.getTerminator()) {
+      curr_wl.push_back(&block);
+      // add block to map with initial entry set
+      block_map.insert({&block, LiveSet()});
+    }
+  }
+}
+
+// Visit the basic blocks in the worklist and update the block_map.
+void LiveSetBlockVisitor::Visit(void) {
+  //bool progress = true;
+  auto currlive = new LiveSet;
+  // if any visit makes progress, continue the loop
+  while (!curr_wl.empty()) {
+    LOG(INFO) << "LSBV: " << curr_wl.size() << " blocks in worklist";
+    //progress = false;
+    for (auto block : curr_wl) {
+      auto newlive = VisitBlock(block, currlive);
+      if (*newlive == *currlive) {
```

Comment:
This is premature. What you are saying here is: "if any block didn't make progress, then the whole system has converged", yet that is too simplistic. It's more that, if the new and current live sets *don't* match, then we want to add the predecessors of a block to the worklist.